### ISSUE TITLE
Allow an option to view mli-constrained types

### ISSUE THREAD

**Issue by bluddy** [ocaml/merlin#706]
It's common practice to constrain the types of values and function in an associated mli file. This works reasonably well for programming, but not so well for merlin. An example:

    A.ml:
    let foo x y = x

   A.mli:
   val foo: int -> int -> int

When working in A.ml, foo will have type `'a -> 'b ->'a`. But we've expressed to the type system that any external call to foo must have constrained type `int -> int -> int`. It would be really be nice to have an 'external type-of' command that would provide the latter, since that is quite likely the information we'd like to have available when programming.

**Comment by rgrinberg** [comment]
> since that is quite likely the information we'd like to have available when programming.

Quite likely, but not always. It does happen that sometimes we restrict the polymorphism in the mli but use the polymorphic version in the implementation. What about just jumping to the declaration of the value in the .mli?

**Comment by bluddy** [comment]
I mean, that's better than nothing, but the problem is that I think what I'm describing is common practice. When you compile a function that's used externally, it's that .mli restriction that counts. It's also common not to restrict the type locally to the .ml file, because the syntax for that is more awkward, but to make that restriction in the .mli file. This is important type information when writing code -- you need to know how external functions view the type of the function -- otherwise you're not getting correct information about it from the perspective you need -- you have to go back to an external file and query that function call to get the correct type as seen from the outside. That's why I would like this as an added type information option.

**Comment by rgrinberg** [comment]
Yeah, I think you're describing a valid problem but the solution you're suggesting doesn't sound like a huge improvement to warrant the complexity.

What you suggest:

1. Ask merlin for the type
2. Switch to the user written type in the mli

What I suggest:

1. Locate the value in the .ml
2. Switch to the value in the mli
2. Go back to where you were editing.

Currently, my approach is a little longer because one has to find the corresponding declaration in the mli. I think if this was automatic, the 2 methods would be comparable. Now given the approach that I describe is already supported by frontends that use LSP, it seems like it would be the easiest thing to support in practice. If you think that switching to a different file is bothersome, in LSP there's a "preview" functionality that saves you this effort.

Could you describe your suggestion a bit more clearly and perhaps give an example of why its advantageous?
